### PR TITLE
Introduce AssertJ for better test code readability

### PR DESCRIPTION
**Pull request by nastra**
I've had really good experiences in the past with [AssertJ](https://assertj.github.io/doc/) when it comes to testing as it's very expressive and makes the code easy to read.

I figured that I'll introduce it and rewrite a bunch of Test classes to show how it makes certain things easier to test.

---
This change is [Reviewable](https://reviewable.io/reviews/projectnessie/nessie/1271)

### CODE REVIEW

**Comment by laurentgo** [comment]
How does it compare with hamcrest (which we already have as a dependency?)

File: clients/client/pom.xml [hunk 1]
```diff
@@ -102,6 +102,12 @@
       <version>1.68</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.assertj</groupId>
+      <artifactId>assertj-core</artifactId>
+      <version>${assertj.version}</version>
```
Comment (snazy) [inline clients/client/pom.xml:108]: can `<version>` be removed here?

File: servers/quarkus-server/src/test/java/org/projectnessie/server/TestRest.java [hunk 2]
```diff
@@ -167,23 +159,25 @@ void referenceNames(String refNamePart) throws NessieNotFoundException, NessieCo
...
     Map<String, Reference> references = tree.getAllReferences().stream()
         .filter(r -> "main".equals(r.getName()) || r.getName().endsWith(refNamePart))
         .collect(Collectors.toMap(Reference::getName, Function.identity()));
-    assertThat(references.values(), containsInAnyOrder(
-        referenceWithNameAndType("main", Branch.class),
-        referenceWithNameAndType(tagName, Tag.class),
-        referenceWithNameAndType(branchName, Branch.class),
-        referenceWithNameAndType(branchName2, Branch.class)));
+
+    assertThat(references).containsAllEntriesOf(
```
Comment (snazy) [inline servers/quarkus-server/src/test/java/org/projectnessie/server/TestRest.java:176]: Looks like this one removes the assertions on the type ("instanceof")

**Review by snazy — CHANGES_REQUESTED** [review]
I think, we can then completely remove the hamcrest-dependency from the top-level `pom.xml`, if rest-assured is fine with that change.

File: servers/quarkus-server/src/test/java/org/projectnessie/server/TestRest.java (hunk 2 above)
Comment (nastra) [reply servers/quarkus-server/src/test/java/org/projectnessie/server/TestRest.java:176]: fixed

File: clients/client/pom.xml (hunk 1 above)
Comment (nastra) [reply clients/client/pom.xml:108]: yep, fixed

**Comment by nastra** [comment]
> How does it compare with hamcrest (which we already have as a dependency?)

@laurentgo Writing AssertJ checks seems to be just easier and more readable when comparing to Hamcrest, as those require nested methods. Also checking exceptional code seems to be much easier with AssertJ than with Junit/Hamcrest.
AssertJ Dev seems to be quite active when looking at https://github.com/assertj/assertj-core/releases, whereas Hamcrest's last release was in Oct 2019.

**Review by snazy — DISMISSED** [review]
LGTM

**Review by snazy — APPROVED** [review]

**Review by laurentgo — APPROVED** [review]

**Pull request merged**